File: php/watch_file.php
#!/usr/bin/env php
<?php

# Check if an argument is provided
if ($argc < 2) {
    echo "Usage: just watch <filename>\n";
    exit(1);
}

# Get the current modification time of the file
$filename = $argv[1] . ".php";
$full_path = getcwd() . DIRECTORY_SEPARATOR . 'src' . DIRECTORY_SEPARATOR . $filename;
if (file_exists($full_path)) {
    echo "Watching file $filename\n\n";
    $current_mod_time = filemtime($full_path);
} else {
    echo "The file $filename does not exist\n";
    exit(1);
}

while (true) {
    clearstatcache(); # Clear cached results of filemtime
    # If the modification time has changed, run the command
    if ($current_mod_time != filemtime($full_path)) {
        $current_mod_time = filemtime($full_path);
        $output = null;
        $retval = null;
        $filename = explode('_', $filename)[0];
        exec("vendor\bin\pest --colors=always --filter={$filename}", $output, $retval);
        echo implode("\n", $output);
    }
    # Wait for a while before checking again
    sleep(1);
}

File: python/watch_file.py
import time
import subprocess
import os
import sys

filename = sys.argv[1]+'.py'

justfile_dir = os.path.dirname(os.path.realpath(__file__))
file_path = os.path.join(justfile_dir, 'src', filename)

# Last change time
current_mod_time = os.stat(file_path).st_mtime

while True:
    # If there was a change run pytest
    if os.stat(file_path).st_mtime != current_mod_time:
        current_mod_time = os.stat(file_path).st_mtime
        subprocess.run(
            f"pytest -s src/__tests__/test_{filename}", shell=True)

    # Wait before checking again
    time.sleep(1)
